packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -738,6 +777,32 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
       // left responses to handle here.
       throwBadResponse(stateP, res);
     }
+  } else if (stateP.controlState === 'READY_TO_REINDEX_SYNC') {
```

Comment:
If we need to synchronise migrators within a Kibana instance, how does the algo for moving documents to different indices behave when there are multiple Kibanas (where no realtime sync between them exists)?

Reply:
That's a very good question.

If we consider the current "single" index scenario, we can already have multiple instances trying to reindex into the same _temporary_ index at the same time.

- Some migrators might override other instances work, but eventually they will all finish (no sync required between them).
- Actually, the first migrator that finishes reindexing into the temporary index will **write-lock** that index and then proceed to clone to the final index. Other migrators that are still bulk indexing will detect the lock and proceed to the next step, also attempting to clone, etc.

Now, regarding the multiple migrators per instance, multiple instances scenario (which BTW will not happen in serverless):

- Instance 1 starts migrators for indices A, B and C. We are moving SO type T from A to B, so A and B need to synchronise their paths.
- Instance 2 does the same thing, a tad later.
- The synchronisation happens twice during the flow:
   - Right after the temp indices for A and B have been created.
     - Imagine the temp indices are created by 1A and 2B.
     - The other migrators (1B, 2A) will eventually attempt to create the temp indices too, detect they already exist and proceed to synchronise (1B will synchronise with 1A, and 2A will sync with 2B).
        - No migrator will start reindexing into temp index until all temp indices have been created ✅ .
  - Right after the temp indices have been filled, and before locking them.
     - Instance 1 will not try to lock temp indices until it is sure that they have all been filled. Same for instance 2.
     - Imaine 1A and 1B have both finished, so they synchronise and lock their indices. They then proceed to clone.
     - 2A has finished writing to temp but 2B is still bulk indexing.
        - 2B will detect the lock and assume another instance has completed indexing to temp.
        - 2A and 2B will synchronise and attempt to lock, attempt to clone, etc.

TLDR: The important thing about the synchronisation points is making sure that all migrators of an instance reach a certain state before proceeding with the flow. Whether 1A reaches the state cause it has completed the work (or because 2A did it faster) does not make a difference, the important is that the work is done.

Reply:
Yes, that does make sense! Thank you for the detailed explanation :D

Your explanation is so useful I think it is worth capturing somewhere in a comment or a MD file :) WDYT?

Reply:
Referenced in https://docs.google.com/document/d/17L_JBIItH-amMMzj455mkM4heZKc7ugb3P4klxzGJeI/edit?usp=sharing